src/Http/Http.Extensions/src/ProblemDetailsJsonOptionsSetup.cs

```diff
@@ -11,23 +11,27 @@ internal sealed class ProblemDetailsJsonOptionsSetup : IPostConfigureOptions<Jso
 {
     public void PostConfigure(string? name, JsonOptions options)
     {
-        switch (options.SerializerOptions.TypeInfoResolver)
+        var typeInfoResolverChain = options.SerializerOptions.TypeInfoResolverChain;
+        if (typeInfoResolverChain.Count == 0)
         {
-            case DefaultJsonTypeInfoResolver:
-                // In this case, the current configuration is using a reflection-based resolver
-                // and we are prepending our internal problem details context to be evaluated
-                // first.
-                options.SerializerOptions.TypeInfoResolver = JsonTypeInfoResolver.Combine(ProblemDetailsJsonContext.Default, options.SerializerOptions.TypeInfoResolver);
-                break;
-            case not null:
-                // Combine the current resolver with our internal problem details context (adding last)
-                options.SerializerOptions.AddContext<ProblemDetailsJsonContext>();
-                break;
-            default:
-                // Not adding our source gen context when TypeInfoResolver == null
-                // since adding it will skip the reflection-based resolver and potentially
-                // cause unexpected serialization problems
-                break;
+            // Not adding our source gen context when the TypeInfoResolverChain is empty,
+            // since adding it will skip the reflection-based resolver and potentially
+            // cause unexpected serialization problems
+            return;
+        }
+
+        var lastResolver = typeInfoResolverChain[typeInfoResolverChain.Count - 1];
+        if (lastResolver is DefaultJsonTypeInfoResolver)
+        {
+            // In this case, the current configuration has a reflection-based resolver at the end
+            // and we are inserting our internal ProblemDetails context to be evaluated
+            // just before the reflection-based resolver.
+            typeInfoResolverChain.Insert(typeInfoResolverChain.Count - 1, ProblemDetailsJsonContext.Default);
```

Comment:
Shouldn't this method account for the possibility that `JsonSerializerOptions` has been made read-only by the user? I believe @brunolins16 was using the copy constructor elsewhere to guard against that case.

Reply:
Use of the copy constructor was added in #45906 to work around https://github.com/dotnet/runtime/issues/80527. See the conversation [here](https://github.com/dotnet/aspnetcore/pull/45906/commits/2540eb41a7329ba0c378e6327ad13fb31a135e55#r1068458691).

That was undone in #46716 once the fix for https://github.com/dotnet/runtime/issues/80527 came in. See https://github.com/dotnet/aspnetcore/issues/46143.

The user shouldn't make the options read-only until all the configuration has run. Them trying to do so, and then us modifying it anyways, isn't really a good solution IMO. Note that the only way they can make the options read-only is if they make it read-only in a `Configure<JsonOptions>()` callback - there's no way for them to access this object earlier. Making it read-only in the middle of configuration doesn't make sense because other configuration needs to run.